pkg/cmd/openshift-operators/util/resourceapply/apps.go

```diff
@@ -0,0 +1,36 @@
+package resourceapply
+
+import (
+	appsv1 "k8s.io/api/apps/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	appsclientv1 "k8s.io/client-go/kubernetes/typed/apps/v1"
+
+	"github.com/openshift/origin/pkg/cmd/openshift-operators/util/resourcemerge"
+)
+
+// ApplyDeployment merges objectmeta and requires matching generation
+func ApplyDeployment(client appsclientv1.DeploymentsGetter, required *appsv1.Deployment, expectedGeneration int64) (*appsv1.Deployment, bool, error) {
+	existing, err := client.Deployments(required.Namespace).Get(required.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		actual, err := client.Deployments(required.Namespace).Create(required)
+		return actual, true, err
+	}
+	if err != nil {
+		return nil, false, err
+	}
+
+	modified := resourcemerge.BoolPtr(false)
+	resourcemerge.EnsureObjectMeta(modified, &existing.ObjectMeta, required.ObjectMeta)
+	if *modified {
+		actual, err := client.Deployments(required.Namespace).Update(existing)
+		return actual, true, err
+	}
+	if existing.ObjectMeta.Generation == expectedGeneration {
+		return existing, false, nil
+	}
+	existing.Spec = *required.Spec.DeepCopy()
+
+	actual, err := client.Deployments(required.Namespace).Update(existing)
```

Comment:
where do we handle conflicts? does the controller just re-apply on next sync?

Reply:
> where do we handle conflicts? does the controller just re-apply on next sync?

Yes, and it is logged and put into the CRD status.